src/neo-vm/Types/Map.cs

```diff
@@ -1,66 +1,60 @@
 using System.Collections;
 using System.Collections.Generic;
 using System.Diagnostics;
-using System.Runtime.CompilerServices;
 
 namespace Neo.VM.Types
 {
     [DebuggerDisplay("Type={GetType().Name}, Count={Count}")]
-    public class Map : CompoundType, IDictionary<PrimitiveType, StackItem>
+    public class Map : CompoundType, IReadOnlyDictionary<PrimitiveType, StackItem>
     {
         private readonly Dictionary<PrimitiveType, StackItem> dictionary;
 
         public StackItem this[PrimitiveType key]
         {
-            [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            get => dictionary[key];
-            [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            set => dictionary[key] = value;
+            get
+            {
+                return dictionary[key];
+            }
+            set
+            {
+                if (ReferenceCounter != null)
+                {
+                    if (dictionary.TryGetValue(key, out StackItem old_value))
+                        ReferenceCounter.RemoveReference(old_value, this);
+                    else
+                        ReferenceCounter.AddReference(key, this);
+                    ReferenceCounter.AddReference(value, this);
+                }
+                dictionary[key] = value;
+            }
         }
 
-        public ICollection<PrimitiveType> Keys => dictionary.Keys;
-        public ICollection<StackItem> Values => dictionary.Values;
+        public IEnumerable<PrimitiveType> Keys => dictionary.Keys;
+        public IEnumerable<StackItem> Values => dictionary.Values;
         public override int Count => dictionary.Count;
-        public bool IsReadOnly => false;
 
-        public Map() : this(new Dictionary<PrimitiveType, StackItem>()) { }
-
-        public Map(Dictionary<PrimitiveType, StackItem> value)
-        {
-            dictionary = value;
-        }
-
-        public void Add(PrimitiveType key, StackItem value)
-        {
-            dictionary.Add(key, value);
-        }
-
-        void ICollection<KeyValuePair<PrimitiveType, StackItem>>.Add(KeyValuePair<PrimitiveType, StackItem> item)
-        {
-            dictionary.Add(item.Key, item.Value);
-        }
-
-        public override void Clear()
+        public Map(Dictionary<PrimitiveType, StackItem> value = null)
+            : this(null, value)
         {
-            dictionary.Clear();
         }
 
-        bool ICollection<KeyValuePair<PrimitiveType, StackItem>>.Contains(KeyValuePair<PrimitiveType, StackItem> item)
+        public Map(ReferenceCounter referenceCounter, Dictionary<PrimitiveType, StackItem> value = null)
+            : base(referenceCounter)
         {
-            return dictionary.ContainsKey(item.Key);
+            dictionary = value ?? new Dictionary<PrimitiveType, StackItem>();
+            if (referenceCounter != null)
+                foreach (var pair in dictionary)
+                {
+                    referenceCounter.AddReference(pair.Key, this);
```

Comment:
Maybe we can have a method for add at same time two references `(this, pair.Key, pair.Value);`

Reply:
What's the difference?